src/vpn-api/worker.ts

```diff
@@ -71,91 +71,245 @@ const getInstanceSubnet = (instanceId: number) => {
 };
 
 const worker = (instanceId: number) => {
-	logger.info(`worker-${instanceId} process started with pid ${process.pid}`);
+	const logger = getLogger('vpn', instanceId);
+	const clientLog: { [key: string]: number } = {};
+
+	logger.notice(`process started with pid=${process.pid}`);
+
+	const fatalErrorHandler = (err: Error) => {
+		logger.emerg(err.message);
+		process.exitCode = 1;
+	};
+
+	const logStateUpdate = (state: clients.DeviceState) => {
+		let stateMsg = `common_name=${state.common_name} connected=${
+			state.connected
+		}`;
+		if (state.virtual_address != null) {
+			stateMsg = `${stateMsg} virtual_address=${state.virtual_address}`;
+		}
+		logger.debug(`successfully updated state for device: ${stateMsg}`);
+	};
+
+	let verbose = VPN_VERBOSE_LOGS;
+	process.on('message', msg => {
+		if (msg === 'toggleVerbosity') {
+			verbose = !verbose;
+			logger.notice(`verbose logging ${verbose ? 'enabled' : 'disabled'}`);
+		}
+	});
 
 	const vpnPort = VPN_BASE_PORT + instanceId;
 	const mgtPort = VPN_BASE_MANAGEMENT_PORT + instanceId;
-	const apiPort = VPN_API_BASE_PORT + instanceId;
-
-	const subnet = getInstanceSubnet(instanceId);
-	const gateway = BALENA_VPN_GATEWAY || subnet.first;
-	const command = [
-		'/usr/sbin/openvpn',
-		'--status',
-		`/run/openvpn/server-${instanceId}.status`,
-		'10',
-		'--cd',
-		'/etc/openvpn',
-		'--config',
-		'/etc/openvpn/server.conf',
-		'--verb',
-		'2',
-		'--dev',
-		`tun${instanceId}`,
-		'--port',
-		`${vpnPort}`,
-		'--management',
-		'127.0.0.1',
-		`${mgtPort}`,
-		'--ifconfig',
-		gateway,
-		subnet.second,
-		'--ifconfig-pool',
-		subnet.third,
-		subnet.last,
-		'--route',
-		subnet.base,
-		subnet.mask,
-		'--push',
-		`route ${gateway}`,
-		'--auth-user-pass-verify',
-		`scripts/auth.sh ${instanceId}`,
-		'via-env',
-		'--client-connect',
-		`scripts/client-connect.sh ${instanceId}`,
-		'--client-disconnect',
-		`scripts/client-disconnect.sh ${instanceId}`,
-	];
-	const openvpn = new forever.Monitor(command, {
-		uid: `openvpn_${instanceId}`,
-		env: process.env,
-		max: 10,
-		spinSleepTime: 1000,
-	}).on('exit', err => {
-		logger.error(`OpenVPN error: ${err.message}`);
-		captureException(err, 'OpenVPN Error');
-		process.exit(2);
+
+	const vpn = new VpnManager(
+		instanceId,
+		vpnPort,
+		mgtPort,
+		getInstanceSubnet(instanceId),
+		BALENA_VPN_GATEWAY,
+	);
+
+	vpn.on('process:exit', (code, signal) => {
+		let msg = 'process exited';
+		if (code != null) {
+			msg = `process exited with code ${code}`;
+		}
+		if (signal != null) {
+			msg = `process terminated with signal ${signal}`;
+		}
+		fatalErrorHandler(new Error(msg));
+	});
+	vpn.on('process:error', fatalErrorHandler);
+
+	// capture openvpn logs
+	vpn.on('log', (level, message) => {
+		// only log warnings (or more severe) unless verbose=true
+		if (verbose || logger.levels[level] <= logger.levels.warning) {
```

Comment:
By setting the transport log level in Winston? This line is only filtering excessively verbose messages from openvpn, there are still notice/info/debug level messages from the application that we want to display (even in non-verbose mode).

Reply:
Ah, so there's a different level for the application vs openvpn?  It still might be worth looking into having winston separate the log levels for the two but this `if` makes sense now.  As an additional thing though it might be worth having `verbose` be a scale rather than a boolean, ie `logger.levels[level] <= verbosity` - but that can wait for a separate PR if at all